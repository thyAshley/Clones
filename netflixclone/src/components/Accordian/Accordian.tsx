import React, { createContext, useState, useContext } from "react";

import { Container, Inner, Title, Frame, Item, Header, Body } from "./styles";
import { AccordianComposition } from "./accordian.types";

interface toggleContextProps {
  toggleShow?: boolean;
  setToggle?: React.Dispatch<React.SetStateAction<boolean>>;
}
const toggleContext = createContext<toggleContextProps>({});

const Accordian: React.FC & AccordianComposition = ({
  children,
  ...restProps
}) => {
  return (
    <Container {...restProps}>
      <Inner>{children}</Inner>
    </Container>
  );
};

Accordian.Title = ({ children, ...restProps }) => {
  return <Title {...restProps}>{children}</Title>;
};

Accordian.Frame = ({ children, ...restProps }) => {
  return <Frame {...restProps}>{children}</Frame>;
};

Accordian.Item = function AccordianItem({ children, ...restProps }) {
  const [toggleShow, setToggle] = useState(false);
  return (
    <toggleContext.Provider value={{ toggleShow, setToggle }}>
      <Item {...restProps}>{children}</Item>
    </toggleContext.Provider>
  );
};
Accordian.Header = function AccordianItem({ children, ...restProps }) {
  const { setToggle } = useContext(toggleContext);
  if (!setToggle) return null;
  return (
    <Header
      onClick={() => setToggle((toggleShow) => !toggleShow)}
      {...restProps}
    >
      {children}
    </Header>
  );
};

Accordian.Body = function AccordianBody({ children, ...restProps }) {
  const { toggleShow } = useContext(toggleContext);
  return toggleShow ? <Body {...restProps}>{children}</Body> : null;
};

export default Accordian;
