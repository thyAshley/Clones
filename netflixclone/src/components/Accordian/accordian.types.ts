export interface AccordianComposition {
  Title: React.FC;
  Frame: React.FC;
  Item: React.FC;
  Header: React.FC;
  Body: React.FC;
}
