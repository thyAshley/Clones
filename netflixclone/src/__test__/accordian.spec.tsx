import React from "react";
import { render, fireEvent, queryByText } from "@testing-library/react";
import faqData from "../fixtures/faqs.json";
import Accordian from "../components/Accordian";

describe("<Accordian />", () => {
  it("renders the <Accordian />", () => {
    const { container, getByText } = render(
      <Accordian>
        <Accordian.Title>Frequently Asked Questions</Accordian.Title>
        <Accordian.Frame>
          {faqData.map((faq) => (
            <Accordian.Item>
              <Accordian.Header>{faq.header}</Accordian.Header>
              <Accordian.Body>{faq.body}</Accordian.Body>
            </Accordian.Item>
          ))}
        </Accordian.Frame>
      </Accordian>
    );

    expect(getByText(/Frequently/gi)).toBeTruthy();
    expect(getByText(/What is Netflix/gi)).toBeTruthy();
    expect(container.firstChild).toMatchSnapshot();
  });

  it("open and closes the <Accordian /> Component", () => {
    const { container, queryByText, debug } = render(
      <Accordian>
        <Accordian.Title>Frequently Asked Questions</Accordian.Title>
        <Accordian.Frame>
          {faqData.map((faq) => (
            <Accordian.Item>
              <Accordian.Header>{faq.header}</Accordian.Header>
              <Accordian.Body>{faq.body}</Accordian.Body>
            </Accordian.Item>
          ))}
        </Accordian.Frame>
      </Accordian>
    );

    expect(queryByText(/netflix is a streaming/gi)).toBeFalsy();
    fireEvent.click(queryByText(/What is net/gi)!);
    expect(queryByText(/netflix is a streaming/gi)).toBeTruthy();
    fireEvent.click(queryByText(/What is net/gi)!);
    expect(queryByText(/netflix is a streaming/gi)).toBeFalsy();
  });
});
