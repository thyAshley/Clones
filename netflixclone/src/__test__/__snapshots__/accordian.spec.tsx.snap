// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Accordian /> renders the <Accordian /> 1`] = `
<div
  class="styles__Container-x1xpw-0 RoLzv"
>
  <div
    class="styles__Inner-x1xpw-2 kmtliL"
  >
    <h1
      class="styles__Title-x1xpw-3 guDAlM"
    >
      Frequently Asked Questions
    </h1>
    <div
      class="styles__Frame-x1xpw-1 enjRfy"
    >
      <div
        class="styles__Item-x1xpw-4 dzuLjK"
      >
        <div
          class="styles__Header-x1xpw-5 eKgNhS"
        >
          What is Netflix?
        </div>
        <img
          alt="Open"
          src="/images/icons/add.png"
        />
      </div>
      <div
        class="styles__Item-x1xpw-4 dzuLjK"
      >
        <div
          class="styles__Header-x1xpw-5 eKgNhS"
        >
          How much does Netflix cost?
        </div>
        <img
          alt="Open"
          src="/images/icons/add.png"
        />
      </div>
      <div
        class="styles__Item-x1xpw-4 dzuLjK"
      >
        <div
          class="styles__Header-x1xpw-5 eKgNhS"
        >
          Where can I watch?
        </div>
        <img
          alt="Open"
          src="/images/icons/add.png"
        />
      </div>
      <div
        class="styles__Item-x1xpw-4 dzuLjK"
      >
        <div
          class="styles__Header-x1xpw-5 eKgNhS"
        >
          How do I cancel?
        </div>
        <img
          alt="Open"
          src="/images/icons/add.png"
        />
      </div>
      <div
        class="styles__Item-x1xpw-4 dzuLjK"
      >
        <div
          class="styles__Header-x1xpw-5 eKgNhS"
        >
          What can I watch on Netflix?
        </div>
        <img
          alt="Open"
          src="/images/icons/add.png"
        />
      </div>
    </div>
  </div>
</div>
`;
